File: api/download.py
from http.server import BaseHTTPRequestHandler
import json
import subprocess
import sys
import os
import tempfile


def install_ytdlp():
    subprocess.check_call(
        [sys.executable, "-m", "pip", "install", "-q", "yt-dlp"],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def get_video_info(url):
    try:
        import yt_dlp
    except ImportError:
        install_ytdlp()
        import yt_dlp

    ydl_opts = {
        "quiet": True,
        "no_warnings": True,
        "skip_download": True,
        "noplaylist": True,
    }

    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        info = ydl.extract_info(url, download=False)

    title = info.get("title", "Facebook Video")
    thumbnail = info.get("thumbnail", "")

    formats = info.get("formats", [])
    qualities = []
    seen = set()

    for f in formats:
        if f.get("vcodec") == "none":
            continue
        height = f.get("height")
        ext = f.get("ext", "mp4")
        furl = f.get("url")
        if not furl or not height:
            continue
        label = f"{height}p"
        if label in seen:
            continue
        seen.add(label)
        fsize = f.get("filesize") or f.get("filesize_approx")
        size_str = f"{round(fsize / 1048576, 1)} MB" if fsize else "MP4 Video"
        vcodec = f.get("vcodec", "")
        tbr = f.get("tbr")
        if tbr:
            size_str = f"{round(tbr)} kbps"
        qualities.append({
            "quality": label,
            "size": size_str,
            "url": furl,
            "ext": ext,
        })

    qualities.sort(key=lambda x: int(x["quality"].replace("p", "")), reverse=True)

    if not qualities:
        raise ValueError("No downloadable video formats found. The video may be private.")

    return {"title": title, "thumbnail": thumbnail, "qualities": qualities}


class handler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length)

        try:
            data = json.loads(body)
        except Exception:
            self._json(400, {"error": "Invalid JSON."})
            return

        url = data.get("url", "").strip()

        if not url:
            self._json(400, {"error": "Please provide a Facebook video URL."})
            return

        valid = (
            "facebook.com" in url
            or "fb.watch" in url
            or "fb.com" in url
        )
        if not valid:
            self._json(400, {"error": "Please provide a valid Facebook video URL."})
            return

        try:
            result = get_video_info(url)
            self._json(200, result)
        except ValueError as e:
            self._json(404, {"error": str(e)})
        except Exception as e:
            self._json(500, {"error": f"Could not extract video: {str(e)}"})

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def _json(self, status, payload):
        body = json.dumps(payload).encode()
        self.send_response(status)
        self._cors()
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *args):
        pass

File: download.js
export default async function handler(req, res) {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") return res.status(200).end();
  if (req.method !== "POST") return res.status(405).json({ error: "Method not allowed" });

  const { url, k } = req.body;

  // ── Mode 2: convert (get final download link) ──
  if (k) {
    try {
      const convertRes = await fetch("https://www.y2mate.com/mates/convertV2/index", {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
          Referer: "https://www.y2mate.com/",
        },
        body: new URLSearchParams({ vid: "", k }).toString(),
      });
      const data = await convertRes.json();
      if (data?.dlink) {
        return res.status(200).json({ dlink: data.dlink });
      }
      return res.status(500).json({ error: "Could not get download link. Try again." });
    } catch (err) {
      console.error(err);
      return res.status(500).json({ error: "Convert failed. Please try again." });
    }
  }

  // ── Mode 1: analyze (get qualities list) ──
  if (!url) {
    return res.status(400).json({ error: "Please provide a Facebook video URL." });
  }

  const isValidUrl =
    url.includes("facebook.com") || url.includes("fb.watch") || url.includes("fb.com");

  if (!isValidUrl) {
    return res.status(400).json({ error: "Please provide a valid Facebook video URL." });
  }

  try {
    const analyzeRes = await fetch("https://www.y2mate.com/mates/analyzeV2/ajax", {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
        Referer: "https://www.y2mate.com/",
      },
      body: new URLSearchParams({
        k_query: url,
        k_page: "Facebook",
        hl: "en",
        q_auto: 0,
      }).toString(),
    });

    const data = await analyzeRes.json();

    if (!data || data.status !== "ok") {
      return res.status(500).json({
        error: "Could not extract video. It may be private or region-restricted.",
      });
    }

    const links = data.links?.mp4 || data.links?.video || {};
    const qualities = Object.entries(links).map(([key, val]) => ({
      quality: val.q || key,
      size: val.size || "Unknown",
      k: val.k,
    }));

    if (!qualities.length) {
      return res.status(404).json({
        error: "No downloadable formats found. The video may be private.",
      });
    }

    return res.status(200).json({
      title: data.title || "Facebook Video",
      thumbnail: data.thumbnail || null,
      qualities,
    });
  } catch (err) {
    console.error(err);
    return res.status(500).json({ error: "Something went wrong. Please try again." });
  }
}
